File: mjx/observation.cpp
#include "mjx/observation.h"

#include <google/protobuf/util/json_util.h>
#include <google/protobuf/util/message_differencer.h>

#include <utility>

#include "mjx/internal/observation.h"

namespace mjx {
Observation::Observation(mjxproto::Observation proto)
    : proto_(std::move(proto)) {}

Observation::Observation(const std::string& json) {
  auto status = google::protobuf::util::JsonStringToMessage(json, &proto_);
  assert(status.ok());
}

const mjxproto::Observation& mjx::Observation::proto() const noexcept {
  return proto_;
}

std::string mjx::Observation::ToJson() const noexcept {
  std::string serialized;
  auto status =
      google::protobuf::util::MessageToJsonString(proto_, &serialized);
  assert(status.ok());
  return serialized;
}

bool Observation::operator==(const Observation& other) const noexcept {
  return google::protobuf::util::MessageDifferencer::Equals(proto_,
                                                            other.proto_);
}

bool Observation::operator!=(const Observation& other) const noexcept {
  return !(*this == other);
}

std::vector<float> Observation::ToFeature(
    const std::string& version) const noexcept {
  auto obs = internal::Observation(proto_);
  return obs.ToFeature(version);
}

std::vector<Action> Observation::legal_actions() const noexcept {
  std::vector<Action> actions;
  for (const auto& action_proto : proto_.legal_actions()) {
    actions.emplace_back(Action(action_proto));
  }
  return actions;
}

std::vector<int> Observation::action_mask() const noexcept {
  auto proto_legal_actions = proto_.legal_actions();
  assert(!proto_legal_actions.empty());
  // TODO: remove magic number 181
  auto mask = std::vector<int>(181, 0);
  for (const auto& proto_action : proto_legal_actions) {
    mask[internal::Action::Encode(proto_action)] = 1;
  }
  return mask;
}

Hand Observation::curr_hand() const noexcept {
  return Hand{proto_.private_observation().curr_hand()};
}
}  // namespace mjx

File: pymjx/mjx/observation.py
import json
from typing import List, Optional

import mjxproto
from google.protobuf import json_format
from mjx.visualizer.svg import save_svg

import mjx


class Observation:
    def __init__(self, cpp_obj):
        self._cpp_obj = cpp_obj

    def legal_actions(self) -> List[mjx.Action]:
        return [mjx.Action(cpp_obj) for cpp_obj in self._cpp_obj.legal_actions()]

    def to_json(self) -> str:
        return self._cpp_obj.to_json()

    def to_proto(self) -> mjxproto.Observation:
        json_data = self.to_json()
        return json_format.ParseDict(json.loads(json_data), mjxproto.Observation())

    def save_svg(self, filename: str, view_idx: Optional[int] = None):
        assert filename.endswith(".svg")
        assert view_idx is None or 0 <= view_idx < 4

        observation = self.to_proto()
        save_svg(observation, filename, view_idx)
